File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def serialize(root: Optional[TreeNode]):
            if not root:
                return "N"
            return f"{root.val},{serialize(root.left)},{serialize(root.right)}"

        root_serialized = serialize(root)
        sub_root_serialized = serialize(subRoot)
        return sub_root_serialized in root_serialized

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int dfs(TreeNode* node, int max_val) {
        if (node == nullptr) return 0;

        int is_good = node->val >= max_val ? 1 : 0;
        return (
            is_good + 
            dfs(node->left, max(max_val, node->val)) + 
            dfs(node->right, max(max_val, node->val))
        );
    }

public:
    int goodNodes(TreeNode* root) {
        return dfs(root, numeric_limits<int>::min());
    }
};
